### ISSUE TITLE
Temp Sensor

### ISSUE THREAD

**Issue by MelbourneDeveloper** [WozSoftware/Woz.TEMPer#1]
Hi,

Thanks for using Device.Net. I'm keen to help out wherever possible with projects using Device.net. is there any chance I can get hold of one of these devices so that I can test with one? Are they expensive?

Christian

**Comment by WozSoftware** [comment]
Hi Christian, Good to know I picked the right library that has active development 😊The devices are dirt cheap, I paid about $10 NZD on AliExpress, link provided but there are many sellers that stock them, you might get even cheaper https://www.aliexpress.com/item/32410119928.html?spm=a2g0s.9042311.0.0.27424c4dLh4dBP 

There are I believe 2 versions of the one I have along with more advanced ones with the sensor on a cable and also some that do humidity, none are bank breaking. Just cut V1.0 on that but I can see more comingThe plan is for central DB service, website to display the stats and small services for the machines that have the sensors so I can record house temps.  

I found some python code that talked with the sensor so extracted that core into C# using your lib. There is not much documentation on the sensor so very much feel your way stuff. I have tested the lib with 2 sensors plugged in, as I borrowed an extra one from a friend so know I handle multiple correctly. 

You code made that easy 😊 Hope that helps and feel free to ask anything

**Comment by WozSoftware** [comment]
Hi Christian, 

As a follow up, the best search on ali is “temperature sensor usb” in computer section. These are the other variants 

TEMPer Mk2 https://www.aliexpress.com/item/32869805457.html?spm=a2g0o.productlist.0.0.2d036affH7T1kj 

TEMPerHUMhttps://www.aliexpress.com/item/32345106214.html?spm=a2g0o.productlist.0.0.2d036affH7T1kj 

Warren

**Comment by MelbourneDeveloper** [comment]
@WozSoftware Hi. I finally put one of these on order for lockdown testing. I had a heap of trouble ordering one through Aliexpress but eventually ordered one through eBay. Hope it arrives soon. I'll let you know when I've got some integration tests in the codebase.

**Comment by WozSoftware** [comment]
@MelbourneDeveloper Nice. I could see in the current state of things orders from Ali could be an issue.

Been a bit distracted on this at the mo but plan to get back to it. This code does appear stable at the mo, so you can use as a test bed if needed. Feel free to shout if you need any more info. But there are a feel it out device as limited info :)

**Comment by MelbourneDeveloper** [comment]
Hi @WozSoftware. I ended up getting the device a while back and have included it as part of some integration tests and samples in the develop branch. Your code helped me to understand how it works. Thanks.

I would like to invite you to the [Device.Net Discord channel](https://discord.gg/ZcvXARm). It would be very helpful to have people around that have implemented the software before. 

This is the [integration test](https://github.com/MelbourneDeveloper/Device.Net/blob/0bf1a528298949006732b7b6d80af63c925d00e0/src/Device.Net.UnitTests/IntegrationTests.cs#L73). It's one of the dongles I will plug in and test every time I do a release.

A [mention of your code here](https://github.com/MelbourneDeveloper/Device.Net/blob/0bf1a528298949006732b7b6d80af63c925d00e0/src/Usb.Net.WindowsSample/Temperature/TemperatureMonitor.cs#L63).

**Comment by WozSoftware** [comment]
Nice. I like the IObserver route, might pick that up later when I get back to this as cleaner way to handle it. I also thought about using Linq style RX extensions as that is an enumerable stream in time for push. Gives you nice query to group/filter streams over time. 

Thanks for the link back :)

 Been a bit distracted with ML.Net and looking at what that is about :)

**Comment by MelbourneDeveloper** [comment]
@WozSoftware exactly. I've been looking at ways to implement the rx extensions. Maybe I can bounce some ideas off you. Want to pick this up on discord?

**Comment by WozSoftware** [comment]
Busy today but can sort something soon if you want.

I have not used RX an anger yet, more explored what it does. From my understanding it is callbacks wrapped up as an Enumerable when they happen. So IEnumerableObserverable

Mouse event are a good example, for this the Enumerable carries the X/Y of the mouse as/when it moves. Linq to objects is a pull semantic in that you ask a question and get the results while RX is the inverse and hence push. So you can smooth the mouse based on what has happened in the last second etc

The key is the sensor becomes a push instead of pull. You push temps at regular intervals, say once per second/minute, and the Linq on this fires as the events come in. You can then filter and manipulate in the same way as Linq to reshape the stream of one per second to one per second but average the last 5 results or throttle to every 5th. 

Eric Meijer does a great talk on the fundamentals, he is funny with it also :) He is one of the people responsible for Linq in C#

https://www.youtube.com/watch?v=sTSQlYX5DU0

**Comment by MelbourneDeveloper** [comment]
@WozSoftware sounds like there's a good discussion here. I've been looking into it and can talk to you about what I experienced. But, let's save it for Discord. 

Thanks, will try to watch the vid.

**Comment by WozSoftware** [comment]
Hi, flicked you my discord details via your website email form so you can connect to keep them off here :)

I don't run a server normally as my bandwidth is low as rural NZ, but we can figure that once we sort some times

Have good day and catch you later

**Comment by MelbourneDeveloper** [comment]
@WozSoftware 
I set up a server specifically for this: https://discord.gg/ZcvXARm

Catch you later.

**Comment by WozSoftware** [comment]
I hate discord sometimes, can you flick new invite. The last one didn't connect to my account but tried to force a new user lol

**Comment by MelbourneDeveloper** [comment]
@WozSoftware sorry I can't. I've posted that invite everywhere.

Yeah, Discord is very confusing. Don't worry about it.

**Comment by WozSoftware** [comment]
If you delete me from the channel I will try to reconnect. Right better get on but l8rs

**Comment by MelbourneDeveloper** [comment]
You mean kick?

![image](https://user-images.githubusercontent.com/16697547/93007781-1692f380-f5b0-11ea-8e9e-c63a20a26a55.png)

**Comment by WozSoftware** [comment]
yep, if you kick might let me reconnect. For some reason clicked the link but then created new acct that needed to be claimed. Which I can't do as my email already used. I will then reconnect from within the client instead of web link
